janusgraph-core/src/main/java/org/janusgraph/core/ConfiguredGraphFactory.java

```diff
@@ -152,13 +153,62 @@ public static JanusGraph close(String graphName) throws Exception {
      * @param graphName String graphName. Corresponding graph can be open or closed.
-     * @throws BackendException If an error occurs during deletion
-     * @throws ConfigurationManagementGraphNotEnabledException If ConfigurationManagementGraph not
      */
-    public static void drop(String graphName) throws Exception {
-        final StandardJanusGraph graph = (StandardJanusGraph) ConfiguredGraphFactory.close(graphName);
-        JanusGraphFactory.drop(graph);
-        removeConfiguration(graphName);
+    public static void drop(String graphName) {
+        final JanusGraphManager jgm = JanusGraphManagerUtility.getInstance();
+        Preconditions.checkNotNull(jgm, JANUS_GRAPH_MANAGER_EXPECTED_STATE_MSG);
+
+        final StandardJanusGraph graph = (StandardJanusGraph) jgm.getGraph(graphName);
+        // Evict the graph from the cache in all nodes in the cluster
+        DropGraphOnEvictionTrigger dropTrigger = new DropGraphOnEvictionTrigger(graphName, graph, getConfigGraphManagementInstance());
+        removeGraphFromCache(graph, dropTrigger);
+        // Block the current thread until the drop operation finish
+        dropTrigger.waitUntilDropFinish();
+    }
+
+    /**
+     * This trigger will fire when the give graph was evicted from all the nodes in the cluster.
+     */
+    private static class DropGraphOnEvictionTrigger implements Callable<Boolean> {
+        private static final Logger log = LoggerFactory.getLogger(DropGraphOnEvictionTrigger.class);
+        private final String graphName;
+        private final JanusGraph graph;
+        private final ConfigurationManagementGraph configManagementGraph;
+        private boolean dropFinished;
+
+        private DropGraphOnEvictionTrigger(String graphName, JanusGraph graph, ConfigurationManagementGraph configManagementGraph) {
+            this.graphName = graphName;
+            this.graph = graph;
+            this.configManagementGraph = configManagementGraph;
+            this.dropFinished = false;
+        }
+
+        @Override
+        public synchronized Boolean call() throws BackendException {
+            try {
+                log.info("Graph {} has been removed from the graph cache on every JanusGraph node in the cluster.", graphName);
+                log.warn("Attempting to drop the graph {}.", graphName);
+                configManagementGraph.removeConfiguration(graphName);
+                JanusGraphFactory.drop(graph);
+                log.warn("Graph {} has been dropped.", graphName);
+                return true;
+            } finally {
+                dropFinished = true;
+                notifyAll();
+            }
+        }
+
+        public synchronized void waitUntilDropFinish() {
+            log.debug("Waiting for graph {} to be evicted from the cache in all JanusGraph nodes...", graphName);
+            while (!dropFinished) {
+                try {
+                    wait();
```

Comment:
The blocking mechanism you implemented looks correct to me, but I recommend using CountDownLatch which is simpler and less error-prone.

Reply:
Thanks for the recommendation, I wasn't familiar with the `CountDownLatch`. Now that I'm learning about it I'm not seeing how this would be simpler for this case, given that we are waiting for only one operation to complete.

Could you please elaborate a little on it? Or just point me to the right place so I can dig some more.

Thanks

Reply:
I guess when I say 'simpler' I mean It's simpler for programmers & reviewers to reason about its correctness. For example, I had to check carefully you had "synchronized" keywords in-place and you used a while loop to wait. In terms of lines of code, there might not be much of a difference.

I recommend reading _Item 81: Prefer concurrency utilities to wait and notify_ from _Effective Java, Third Edition_. [Here](http://reader.epubee.com/books/mobile/e0/e0931e78a3cb5e4e6a04116a22bcbded/text00005.html#filepos0001169971) is a resource online. As it suggests,

> Given the difficulty of using wait and notify correctly, you should use the higher-level concurrency utilities instead.

There are some discussions around this on StackOverflow as well, like [this](https://stackoverflow.com/questions/2536692/a-simple-scenario-using-wait-and-notify-in-java) and [this](https://stackoverflow.com/questions/5999100/is-there-a-block-until-condition-becomes-true-function-in-java).

Regarding the usage of CountDownLatch, I am sure you could find tons of examples online. There are also a few usages of it in the JanusGraph codebase.

Reply:
Thanks for the detailed answer, I appreciate your time investment in it and in this review.

I will update this PR during the holidays to include your suggestions.

Reply:
I update this to use `CountDownLatch` instead of _wait/notifyAll_.

I also removed the _while (!dropFinished)_ check because it was causing the Gremlin script engine to lock forever if the script times out before the eviction completes.